Add a cron job to catch-up soft deactivated users

punchagan opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->

This PR adds a cron job to catch-up soft deactivated users.  The cron job can be turned off by turning off the `AUTO_CATCH_UP_SOFT_DEACTIVATED_USERS` flag. 

It also adds pagination to bulk creation of missing UserMessage rows.

timabbott commented on zerver/lib/soft_deactivation.py:
```diff
@@ -158,8 +159,15 @@ def add_missing_messages(user_profile: UserProfile) -> None:
+            user_messages_to_insert[0:BULK_CREATE_BATCH_SIZE],
+            user_messages_to_insert[BULK_CREATE_BATCH_SIZE:])
+        UserMessage.objects.bulk_create(messages)
+        # Update last_active_message_id if there are more messages to create
```
Why don't we do this unconditionally?

punchagan replied:
Yes, we should do this unconditionally. I thought the last update was being done by the callers of this code, but it isn't. In any case, it would be better to do this unconditionally, here.

timabbott commented on zerver/management/commands/soft_deactivate_users.py:
```diff
@@ -76,11 +74,11 @@ def handle(self, *args: Any, **options: str) -> None:
+                    user_emails_found = {user.email for user in users_to_deactivate}
+                    user_emails_not_found = ', '.join(set(user_emails) - user_emails_found)
+                    raise Exception('Users with emails %s were not found. '
+                                    'Check if they are correct.' % (user_emails_not_found))
```
We might want to print the emails one-per line if we thought this would be a common code path with multiple emails, but I think it really isn't given that the default flow is auto-detection.

punchagan replied:
Raising an exception once per invalid email ID would be annoying if there were more than one invalid email IDs supplied - the user would've to run the command multiple times.  

I could change this to print the email IDs one per line.

timabbott replied:
yeah, one per line would be "better" (and we should raise `CommandError` to proper management command styling).  But as I mentioned, I think realistically the use case of more-than-one directly specified email is likely to be very uncommon, so I'm also fine with punting on doing a nicer display.

timabbott commented on zerver/lib/soft_deactivation.py:
```diff
@@ -212,7 +212,7 @@ def do_soft_deactivate_users(users: List[UserProfile]) -> List[UserProfile]:
     return users_soft_deactivated
 
-def maybe_catch_up_soft_deactivated_user(user_profile: UserProfile) -> Union[UserProfile, None]:
+def maybe_activate_soft_deactivated_user(user_profile: UserProfile) -> Union[UserProfile, None]:
```
Maybe an even better name would be `reactivate_user_if_soft_deactivated`?

punchagan replied:
Yes, that sounds better. Changing!

timabbott commented on zerver/tests/test_soft_deactivation.py:
```diff
@@ -90,3 +92,50 @@ def test_do_soft_activate_users(self) -> None:
+            user.refresh_from_db()
+            self.assertTrue(user.long_term_idle)
+            catch_up_received = UserMessage.objects.filter(message_id=message_id).count()
+            self.assertEqual(already_received + len(users), catch_up_received)
```
We should validate the value of `last_active_message_id` here as well, since bugs in management of that seem like a pretty plausible way this algorithm could go wrong.  (and I think it's possible this PR has a bug where we're not doing that properly).

punchagan replied:
Yep, that's a good check to have!

timabbott commented on puppet/zulip/files/cron.d/catch-up-soft-deactivated-users:
```diff
@@ -0,0 +1,6 @@
+USER=zulip
+
+# Cron job to catch-up soft deactivated users daily.
+0 17 * * * zulip /home/zulip/deployments/current/manage.py soft_deactivate_users -c >/dev/null
```
At the very least, I think this should be the same cron file used to soft-deactivate users (and just an additional line in it), but see my comment above that it might be even better to just have this be a part of what the main soft-deactivation cron job does.

punchagan replied:
Yep, having it in the same file is nicer! But, I wasn't sure if that was the way to go. I'll change this.

timabbott commented on zerver/management/commands/soft_deactivate_users.py:
```diff
@@ -65,6 +72,26 @@ def handle(self, *args: Any, **options: str) -> None:
+                users_to_catch_up = UserProfile.objects.filter(
+                    email__in=user_emails,
+                    **filter_kwargs
+                )
```
Should both this code path and the `activate` code path be also filtering for `long_term_idle=True`?  Seems like a valuable perf optimization.

punchagan replied:
Yep! Except, I guess the error handling needs to be a little more careful - and we need to separately print the users who didn't need to be activated or caught up and the email IDs which were invalid.

punchagan replied:
Punting on this, since we don't expect the code path where emails are explicitly supplied to have more than a handful of emails.

timabbott commented:
@punchagan this is a great start!  I posted a batch of comments, but aside from those, I'm very happy with this PR.

punchagan commented:
Thanks for the review, @timabbott!

I'll respond to the comments and update the PR later today/tomorrow!

punchagan commented on zerver/management/commands/soft_deactivate_users.py:
```diff
@@ -65,6 +72,26 @@ def handle(self, *args: Any, **options: str) -> None:
+
+            if users_to_catch_up:
+                users_caught_up = do_catch_up_soft_deactivated_users(users_to_catch_up)
+                logger.info('Caught up %d soft deactivated user(s)' % (len(users_caught_up)))
```
So, as I see it we always want to do the "catching up" also if soft deactivation is being done. May be catch-up should happen by default if deactivate is being run (without any explicit users being specified).  If we don't want catch-up to happen, may be that should be a flag `--no-catch-up` ? :thinking:

timabbott replied:
yeah, that's pretty reasonable.

punchagan commented:
I've updated the PR addressing most of the comments, @timabbott. There's a couple of things that may be need a little more discussion. Thanks again for your super helpful and quick review!

timabbott commented:
I merged a half-dozen commits.  We should probably mock `logging.info` in some of the tests to clean up this output:
```
Running zerver.tests.test_soft_deactivation.UserSoftDeactivationTests.test_do_catch_up_users
2019-03-14 00:23:21.061 INFO [] Soft-deactivated batch of 2 users; 0 remain to process
2019-03-14 00:23:21.112 INFO [] Caught up 2 soft-deactivated users
Running zerver.tests.test_soft_deactivation.UserSoftDeactivationTests.test_do_soft_activate_users
2019-03-14 00:23:21.251 INFO [] Soft-deactivated batch of 3 users; 0 remain to process
Running zerver.tests.test_soft_deactivation.UserSoftDeactivationTests.test_do_soft_deactivate_user
Running zerver.tests.test_soft_deactivation.UserSoftDeactivationTests.test_do_soft_deactivate_users
2019-03-14 00:23:21.336 INFO [] Soft-deactivated batch of 3 users; 0 remain to process
Running zerver.tests.test_soft_deactivation.UserSoftDeactivationTests.test_get_users_for_catch_up
Running zerver.tests.test_soft_deactivation.UserSoftDeactivationTests.test_get_users_for_soft_deactivation
```

Also, after thinking a bit more about the catch-up feature, I think I like the approach you suggested of having the catch-up run automatically after any all-users run of `manage.py soft_deactivate_users`, at least if `settings.AUTO_CATCH_UP_SOFT_DEACTIVATED_USERS` is enabled.  We might want to refactor that management command a bit to put most of the actual business logic in functions (maybe also moving them to `zerver/lib/soft_deactivation.py` and adding tests; as a general rule we try to keep real code out of management commands) so that the flow can be really readable.

And then we don't need to adjust the cron job situation at all.

timabbott commented:
(Also, wanted to say these are really nice cleanups, so thanks for doing them!)

punchagan commented:
@timabbott Thanks for the review and the (partial) merge!  

I updated the last few commits based on your comments. I think I've addressed all the comments, and this is ready for another review.

timabbott commented:
Merged as the commits ending with d75d2c9974339d269ba0add9220f19c7440929d7, after editing the last commit message to close #8858 (since after I thought about it, this does resolve that issue in the default configuration).  We should probably discuss next steps for chat.zulip.org, where we will probably want this option to be off by default.

Thanks @punchagan!

punchagan commented:
Thanks for the edits, and the merge, @timabbott !
